### ISSUE TITLE
Inference of NLI is not deterministic

### ISSUE THREAD

**Issue by takuma-ynd** [uclmr/jack#384]
Even when I feed the same QASetting to a reader, **inference result changes every time**.
It happens both on CPU and GPU.

The following is the result of inferences and scores, performed 30 times on the **same** QASetting object.
```
claim: Edison is genious
evidence: Edison is stupid
contradiction   0.3829
entailment      0.5994
entailment      0.4744
neutral 0.3576
contradiction   0.6101
entailment      0.6437
entailment      0.6857
entailment      0.5381
entailment      0.8286
neutral 0.4443
contradiction   0.6229
entailment      0.8894
entailment      0.4212
entailment      0.5915
neutral 0.4956
entailment      0.5666
entailment      0.7964
contradiction   0.3765
contradiction   0.4539
contradiction   0.7539
```

For reproducing the issue:
- reader: pre-trained dam model ( same as the one in quickstart.ipynb )
- program:
```python
import argparse
from jack import readers
from jack.core import QASetting


if __name__ == "__main__":
    parser = argparse.ArgumentParser("read claim/evidence and output verdict")
    parser.add_argument(
        "--saved_reader", help="path to saved reader directory")
    args = parser.parse_args()


    print("loading reader from file:", args.saved_reader)
    dam_reader = readers.reader_from_file(args.saved_reader)

    claim = "Edison is genious"
    evidence = "Edison is stupid"

    print("claim:", claim)
    print("evidence:", evidence)

    setting = QASetting(question=claim, support=[evidence])
    for i in range(20):
        pred = dam_reader([setting])
        print("{}\t{:.4}".format(pred[0][0].text, pred[0][0].score))
```

**Comment by takuma-ynd** [comment]
Not sure but dropout might be enabled even in the inference step (?)

**Comment by dirkweissenborn** [comment]
@takuma-ynd you are right, that is a particular bug in the DAM reader, can you try ESIM, just to be sure? Sorry for that.

@pminervini could you fix this?

**Comment by takuma-ynd** [comment]
@dirkweissenborn 
I tried it on ESIM and got the same (wrong) result.

**Comment by takuma-ynd** [comment]
I found that it's definitely a dropout issue.
when I tried it with
```python
dam_reader = readers.reader_from_file(args.saved_reader, dropout=0.0)
```
I got 
```
claim: Edison is genious
evidence: Edison is stupid
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
entailment      0.6347
```

**Comment by pminervini** [comment]
I'm on it wrt DAM, Dirk will check the ESIM code!

**Comment by dirkweissenborn** [comment]
We should have a test for that

**Comment by dirkweissenborn** [comment]
@takuma-ynd I pushed a fix for ESIM, can you please check again?

**Comment by takuma-ynd** [comment]
It looks much stabler than before!
```
claim: Edison is genious
evidence: Edison is stupid
neutral 0.9805
neutral 0.8416
neutral 0.9304
neutral 0.9337
neutral 0.9599
neutral 0.8705
neutral 0.927
neutral 0.9294
neutral 0.947
neutral 0.9265
neutral 0.8926
neutral 0.9521
neutral 0.9291
neutral 0.9132
neutral 0.8711
neutral 0.9554
neutral 0.8564
neutral 0.8379
neutral 0.9405
neutral 0.928
```
However, when I fix the dropout explicitly, I could get precisely same score for each line.

**Comment by dirkweissenborn** [comment]
Is this DAM or ESIM?

**Comment by takuma-ynd** [comment]
@dirkweissenborn 
It's ESIM (trained on the previous version)
I said it got stabler, but that was not right.
It looks the result hasn't really changed.

**Comment by dirkweissenborn** [comment]
@takuma-ynd now it is fixed. Sorry for that.

**Comment by takuma-ynd** [comment]
@dirkweissenborn ESIM works fine now. thanks!

**Comment by pminervini** [comment]
@takuma-ynd is DAM fine now?

**Comment by takuma-ynd** [comment]
@pminervini It works fine now! Thanks!
One smaller issue is:
score is fixed in a single run, however, the score changes when I re-run the model.

1st run
```
claim: Edison is genious
evidence: Edison is stupid
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
entailment      0.6156
```

2nd run
```
claim: Edison is genious
evidence: Edison is stupid
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
entailment      0.6373
```

For ESIM, this issue looks already resolved.

One can fix the value by explicitly set the seed for tensorflow and numpy and random, however, it would be nicer if the seed is fixed inside of jack.

**Comment by pminervini** [comment]
@dirkweissenborn what do you think may be the cause?

**Comment by takuma-ynd** [comment]
@dirkweissenborn @pminervini Thanks for the quick work!

**Comment by dirkweissenborn** [comment]
@pminervini I have no idea what that is. Very strange. For ESIM this does not seem to happen.

**Comment by pminervini** [comment]
@dirkweissenborn it was the START token in DAM! It was randomly initialised, but for some reason it was not saved in the model (maybe because it was not trainable).

**Comment by dirkweissenborn** [comment]
Oh, good catch. I think you can either sample using a fixed seed or let it train. However, this means we have to retrain that model, right? :/

**Comment by pminervini** [comment]
Sure, I'll be on it tonight
